In [1]:
from numpy import loadtxt
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import roc_auc_score , classification_report, mean_squared_error, r2_score,accuracy_score,confusion_matrix
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from sklearn.preprocessing import LabelEncoder

In [2]:
fake_real_data = pd.read_csv("fake_job_postings.csv")

In [3]:
fake_real_data.head()

job_id                                      title            location  \
0       1                           Marketing Intern    US, NY, New York   
1       2  Customer Service - Cloud Video Production      NZ, , Auckland   
2       3    Commissioning Machinery Assistant (CMA)       US, IA, Wever   
3       4          Account Executive - Washington DC  US, DC, Washington   
4       5                        Bill Review Manager  US, FL, Fort Worth   

  department salary_range                                    company_profile  \
0  Marketing          NaN  We're Food52, and we've created a groundbreaki...   
1    Success          NaN  90 Seconds, the worlds Cloud Video Production ...   
2        NaN          NaN  Valor Services provides Workforce Solutions th...   
3      Sales          NaN  Our passion for improving quality of life thro...   
4        NaN          NaN  SpotSource Solutions LLC is a Global Human Cap...   

                                         description  \
0  Food52, a fast-growing, James Beard Award-winn...   
1  Organised - Focused - Vibrant - Awesome!Do you...   
2  Our client, located in Houston, is actively se...   
3  THE COMPANY: ESRI – Environmental Systems Rese...   
4  JOB TITLE: Itemization Review ManagerLOCATION:...   

                                        requirements  \
0  Experience with content management systems a m...   
1  What we expect from you:Your key responsibilit...   
2  Implement pre-commissioning and commissioning ...   
3  EDUCATION: Bachelor’s or Master’s in GIS, busi...   
4  QUALIFICATIONS:RN license in the State of Texa...   

                                            benefits  telecommuting  \
0                                                NaN              0   
1  What you will get from usThrough being part of...              0   
2                                                NaN              0   
3  Our culture is anything but corporate—we have ...              0   
4                              Full Benefits Offered              0   

   has_company_logo  has_questions employment_type required_experience  \
0                 1              0           Other          Internship   
1                 1              0       Full-time      Not Applicable   
2                 1              0             NaN                 NaN   
3                 1              0       Full-time    Mid-Senior level   
4                 1              1       Full-time    Mid-Senior level   

  required_education                   industry              function  \
0                NaN                        NaN             Marketing   
1                NaN  Marketing and Advertising      Customer Service   
2                NaN                        NaN                   NaN   
3  Bachelor's Degree          Computer Software                 Sales   
4  Bachelor's Degree     Hospital & Health Care  Health Care Provider   

   fraudulent  
0           0  
1           0  
2           0  
3           0  
4           0

In [4]:
columns = list(fake_real_data.columns)
null_columns = dict()
type_null_columns = list()
#Find all null values in the columns
for values in columns:
    null_values = fake_real_data[values].isnull().sum()
    #Add the to a dictionary for further processing
    if null_values > 0:
        null_columns[values] = null_values
        type_null_columns.append(type(fake_real_data[values][0]))
    print("Number of nan values in %s is %d"%(values,null_values))

Number of nan values in job_id is 0
Number of nan values in title is 0
Number of nan values in location is 346
Number of nan values in department is 11547
Number of nan values in salary_range is 15012
Number of nan values in company_profile is 3308
Number of nan values in description is 1
Number of nan values in requirements is 2695
Number of nan values in benefits is 7210
Number of nan values in telecommuting is 0
Number of nan values in has_company_logo is 0
Number of nan values in has_questions is 0
Number of nan values in employment_type is 3471
Number of nan values in required_experience is 7050
Number of nan values in required_education is 8105
Number of nan values in industry is 4903
Number of nan values in function is 6455
Number of nan values in fraudulent is 0


In [6]:
fake_real_data.describe()

job_id  telecommuting  has_company_logo  has_questions  \
count  17880.000000   17880.000000      17880.000000   17880.000000   
mean    8940.500000       0.042897          0.795302       0.491723   
std     5161.655742       0.202631          0.403492       0.499945   
min        1.000000       0.000000          0.000000       0.000000   
25%     4470.750000       0.000000          1.000000       0.000000   
50%     8940.500000       0.000000          1.000000       0.000000   
75%    13410.250000       0.000000          1.000000       1.000000   
max    17880.000000       1.000000          1.000000       1.000000   

         fraudulent  
count  17880.000000  
mean       0.048434  
std        0.214688  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [8]:
corr_matrix = fake_real_data.corr()
corr_matrix['job_id'].sort_values(ascending=False)

job_id              1.000000
fraudulent          0.079872
telecommuting      -0.004559
has_company_logo   -0.014539
has_questions      -0.087025
Name: job_id, dtype: float64

In [9]:
fake_real_data = fake_real_data.dropna()
fake_real_data = fake_real_data.drop('job_id',axis=1)

In [10]:
#Drop all the null values from rest of columns
fake_real_data = fake_real_data.drop(['company_profile','description','benefits','requirements','salary_range'],axis=1)

In [11]:
cat_var = ['title','location','department','employment_type','required_experience','required_education','industry','function']

le = LabelEncoder()
fake_real_data[cat_var] = fake_real_data[cat_var].apply(lambda x: le.fit_transform(x))

fake_real_data[cat_var]

title  location  department  employment_type  required_experience  \
6        255        10          12                1                    5   
15       596        91         193                1                    3   
23       599       110          37                1                    3   
98       269       100         155                1                    5   
102      346        43         147                1                    2   
134      488        18         196                1                    5   
153      500        17         177                1                    5   
160      213       216          98                1                    0   
179      293        32         171                1                    5   
197      534        64         207                1                    5   
198      322        38          36                1                    2   
215      271       108         155                1                    5   
241      310        32         100                1                    0   
245      322        24          36                1                    2   
247      254        24         146                1                    5   
255      443        24         100                1                    0   
298      476       201          74                1                    6   
314      294        13         118                1                    5   
324      549       113         148                1                    5   
343      430        54         208                1                    5   
371      354       144         142                1                    1   
438      573       150          31                1                    6   
478      309       189           7                1                    2   
527      489        92         142                1                    5   
559      325        92         193                1                    5   
571      149       167          87                1                    1   
612      390       168         159                3                    2   
628      586       106         156                1                    5   
633      264        71          47                0                    2   
690      119       112         142                1                    0   
...      ...       ...         ...              ...                  ...   
16723    602        87         189                1                    0   
16858    495        32         221                1                    0   
16871    161        32         221                1                    0   
16988    520         0         193                1                    2   
16992    508       121         229                1                    5   
16993    576       121         229                1                    5   
17028    289       129          30                1                    0   
17048    451       162         125                1                    5   
17093    219       111         102                1                    2   
17116    387       120         193                1                    0   
17160    518        10         159                1                    0   
17167    570        10         142                1                    2   
17201    539       174         127                1                    5   
17216    302        45         129                1                    4   
17244    174       182         219                1                    1   
17281    555        77         122                1                    6   
17312    604       167         142                1                    2   
17318    197       158         193                1                    6   
17362    141       167         217                1                    0   
17371      9        77          66                1                    6   
17488    341        77          99                0    

In [66]:
#Encoding non numerical attributes to numerical attributes
mapping = {k: v for v, k in enumerate(fake_real_data.title.unique())}
fake_real_data['title'] = fake_real_data.title.map(mapping)
fake_real_data.head()

job_id  title            location  \
1       2      0      NZ, , Auckland   
3       4      1  US, DC, Washington   
4       5      2  US, FL, Fort Worth   
6       7      3      DE, BE, Berlin   
8       9      4   US, FL, Pensacola   

                                     company_profile  \
1  90 Seconds, the worlds Cloud Video Production ...   
3  Our passion for improving quality of life thro...   
4  SpotSource Solutions LLC is a Global Human Cap...   
6  Founded in 2009, the Fonpit AG rose with its i...   
8  Solutions3 is a woman-owned small business who...   

                                         description  \
1  Organised - Focused - Vibrant - Awesome!Do you...   
3  THE COMPANY: ESRI – Environmental Systems Rese...   
4  JOB TITLE: Itemization Review ManagerLOCATION:...   
6  Your Responsibilities: Manage the English-spea...   
8  Implementation/Configuration/Testing/Training ...   

                                        requirements  telecommuting  \
1  What we expect from you:Your key responsibilit...              0   
3  EDUCATION: Bachelor’s or Master’s in GIS, busi...              0   
4  QUALIFICATIONS:RN license in the State of Texa...              0   
6  Your Know-How:                                ...              0   
8  MUST BE A US CITIZEN.An active TS/SCI clearanc...              0   

   has_company_logo  has_questions employment_type  \
1                 1              0       Full-time   
3                 1              0       Full-time   
4                 1              1       Full-time   
6                 1              1       Full-time   
8                 1              1       Full-time   

                              industry  fraudulent  
1            Marketing and Advertising           0  
3                    Computer Software           0  
4               Hospital & Health Care           0  
6                         Online Media           0  
8  Information Technology and Services           0

In [12]:
#Putting feature variable to X (target value)
X = fake_real_data.drop('fraudulent',axis=1)

# Putting response variable to y (Fraudulent column is our target column)
y = fake_real_data['fraudulent']

# Splitting the data into train and test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=101)

In [13]:
rfc = RandomForestClassifier()
rfc.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [14]:
predictions = rfc.predict(X_test)
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       0.95      1.00      0.98       207
           1       1.00      0.62      0.76        26

    accuracy                           0.96       233
   macro avg       0.98      0.81      0.87       233
weighted avg       0.96      0.96      0.95       233



In [15]:
model = XGBClassifier()
model.fit(X_train, y_train)

XGBClassifier(base_score=0.5, booster=None, colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints=None,
              learning_rate=0.300000012, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints=None,
              n_estimators=100, n_jobs=0, num_parallel_tree=1,
              objective='binary:logistic', random_state=0, reg_alpha=0,
              reg_lambda=1, scale_pos_weight=1, subsample=1, tree_method=None,
              validate_parameters=False, verbosity=None)

In [16]:
predictions = model.predict(X_test)
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       0.97      1.00      0.98       207
           1       0.95      0.77      0.85        26

    accuracy                           0.97       233
   macro avg       0.96      0.88      0.92       233
weighted avg       0.97      0.97      0.97       233



# Performance comparision on XGBoost classifier vs Random Forest on Non descriptive columns

## Accuracy for XGBoost = 97%

## Accuracy for Random Forest = 96%